control-plane/agents/src/bin/core/tests/rebuild/mod.rs

```diff
@@ -0,0 +1,260 @@
+            tracing::info!("Pool is Online now");
+            return Some(pool);
+        }
+
+        if std::time::Instant::now() > (start + timeout)
+            && (pool.state.is_none() || pool.state.clone().unwrap().status != PoolStatus::Online)
+        {
+            return None;
+        }
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
+}
+
+// This test
+// 1. creates a volume with 3 replicas, faults one of them by restarting container.
+// 2. Expects child's replica to be online again within 15sec period.
+// 3. Attempts to online the child.
+// 4. Child should get successfully onlined and partial rebuild happen.
+// 5. If online call fails, full rebuild of child happens.
+#[tokio::test]
+async fn replica_wait_and_partial_rebuild() {
+    let cluster = build_cluster(4, 52428800, 0).await; // 50MiB pool size
+
+    let vol_target = cluster.node(0).to_string();
+    let api_client = cluster.rest_v00();
+    let volume_api = api_client.volumes_api();
+
+    let body = CreateVolumeBody::new(VolumePolicy::new(true), 3, 10485760u64, false);
+    let volid = VolumeId::new();
+    let volume = volume_api.put_volume(&volid, body).await.unwrap();
+    let volume = volume_api
+        .put_volume_target(
+            &volume.spec.uuid,
+            PublishVolumeBody::new_all(
+                HashMap::new(),
+                None,
+                vol_target.clone().to_string(),
+                models::VolumeShareProtocol::Nvmf,
+                None,
+                cluster.csi_node(0),
+            ),
+        )
+        .await
+        .unwrap();
+
+    let volume_state = volume.state;
+    assert!(volume_state.status == VolumeStatus::Online.into());
+    let nexus = volume_state.target.unwrap();
+    assert_eq!(nexus.children.len(), 3);
+    nexus
+        .children
+        .iter()
+        .for_each(|c| assert!(c.state == ChildState::Online.into()));
+
+    // Check where the replicas are, apart from vol target node.
+    let replicas = api_client.replicas_api().get_replicas().await.unwrap();
+    tracing::info!({?replicas}, "Available replicas - ");
+    let testrep = replicas.iter().find(|r| r.node != vol_target).unwrap();
+    let testrep_node = &testrep.node;
+    tracing::info!(
+        "Restarting node {testrep_node} having replica {}",
+        testrep.uri.clone()
+    );
+
+    // Stop the container and start again.
+    cluster
+        .composer()
+        .restart(&testrep_node.to_string())
+        .await
+        .expect("container stop failure");
+
+    // Check the pool on restarted node is imported back.
+    // We create one pool per node so pool index is 0.
+    // Can't see any other better way to get node index from name,
+    // so utilizing the container name format io-engine-<num>.
+    let nodeidx: u32 = testrep_node
+        .as_str()
+        .split('-')
+        .last()
+        .expect("Couldn't parse nodeidx")
+        .parse::<u32>()
+        .unwrap()
+        - 1;
```

Comment:
this seems a bit brittle, why do we need to do this?
Can't we just use node-name rather than node index?

Reply:
to get the poolid of a pool on a node via cluster.pool() , we need the node index. Since the io-engine node being restarted is randomly chosen here, I didn't find any other way to get the index of node except utilising the name to derive that.

Reply:
I'm still confused, we have the node and pool in the replica, we can just pass that?

Reply:
ok, I didn't know models Replica had pool String present. This should be ok I guess then. No need of node.

`let pool = api_client.pools_api().get_pool(testrep.pool.as_str()).await.unwrap();`